Support batch table binary section for pnts

lilleyse opened the pull request:
For https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/32#issuecomment-232107248

The pnts format now loads colors through the batch table JSON/binary, or uses a default color if there is no batch table or no `TILES3D_RGB` property. I'm not creating a `batchTableResources` for `Points3DTileContent` since that will probably require some more thought it we even want per-point properties.

One question: do we want to allow the pnts format to specify a default color? If it's in the batch table it could look like:

```javascript
{
    "TILES3D_COLOR" : [1.0, 0.0, 0.0]
}
```

lilleyse commented on Source/Scene/Points3DTileContent.js:
```diff
@@ -166,11 +169,46 @@ define([
+            colors = new Uint8Array(length);
+            for (var i = 0; i < length; ++i) {
+                colors[i] = 64;
+            }
```
Hm I think I'm going to fix this so the attribute is just a single value.

lilleyse replied:
Actually this won't work easily with `GeometryAttribute`.

pjcozzi replied:
Can't we provide a uniform using an appearance?

Otherwise put a `TODO` here (not `PERFORMANCE_IDEA`) so we optimize this before merging into `master`.

lilleyse replied:
I'll try that.

pjcozzi commented:
>  if we even want per-point properties

It doesn't need to be part of this PR and it might be a new class in Cesium, but we will want per-point properties that can be used for dynamic styling and vertex shader generation, see https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/22#issuecomment-231618325

pjcozzi commented:
> One question: do we want to allow the pnts format to specify a default color?

OK...it should include alpha and be the same range as the per-point color, e.g., `[0, 255]` or whatever it ends up being without compression.  No need to finalize now, but it will need to be consistent.

lilleyse commented:
That reminds me I should also include a TILES3D_RGBA semantic.

pjcozzi commented:
> That reminds me I should also include a TILES3D_RGBA semantic.

+1

pjcozzi commented:
Just those comments.

lilleyse commented:
Updated to support RGBA. What is the best way to render `PointGeometry` with alpha blending?

lilleyse commented:
Ah I messed up the merge. Fixing that now.

lilleyse commented:
> Updated to support RGBA. What is the best way to render PointGeometry with alpha blending?

Never mind, I have an idea. It will require modifying the `PointApperance` shaders though, which we'll end up doing anyways.

pjcozzi commented on Source/Scene/Points3DTileContent.js:
```diff
@@ -166,11 +169,46 @@ define([
+            // Get the point colors
+            var tiles3DRGB = batchTableJSON.TILES3D_RGB;
+            var tiles3DRGBA = batchTableJSON.TILES3D_RGBA;
+            if (defined(tiles3DRGB)) {
```
What do you think about https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/22#issuecomment-232489831

Should the tile be invalid (in practice if someone wants multiple per-point colors, they would use their own semantics)?  Or give should RGBA win over RGB?

lilleyse commented:
Updated. I'm building the shaders based on whether it's RGB, RGBA, or constant color, but still using `PointGeometry` and `PointAppearance`. I also added `TILES3D_COLOR` option from here: https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/22#issuecomment-232488235.

pjcozzi commented:
Very nice, thanks @lilleyse!

pjcozzi commented on Source/Scene/Points3DTileContent.js:
```diff
@@ -166,11 +168,97 @@ define([
+        var batchTableBinaryByteLength = view.getUint32(byteOffset, true);
+        byteOffset += sizeOfUint32;
+
+        var positions = new Float32Array(arrayBuffer, byteOffset, pointsLength * 3);
```
Actually, in b3dm and i3dm, the batch table comes before glTF.  We should be consistent here for now and move the positions payload after the batch table; however, we might want to switch them all later if there is a benefit, e.g., to avoid parsing it in some cases?

lilleyse replied:
Ah that's true, I'll make the change in a new PR.

lilleyse replied:
I'm surprised we haven't run into this before, but TypedArray constructors expect their byte offset to be aligned to the type. For two of the points tilesets, the byte length of the batch table is odd and this causes problems when creating a `Float32Array` of the positions. Do we want to require byte alignment in the spec?

pjcozzi replied:
Yes, definitely need to say this in the spec.  @tfili and I have ran into this before and I know some of the work we did included the padding

pjcozzi replied:
glTF spec guarantees this btw. https://github.com/KhronosGroup/glTF/tree/master/specification#bufferview-and-accessor-byte-alignment
